src/main.rs

```diff
@@ -1,245 +1,202 @@
-    };
-
-    println!("Result CID: {}", output);
-
-    network_client.start_providing(closure_cid2).await;
-
-    loop {
-        match network_events.next().await {
-            Some(network::Event::InboundRequest { request, channel }) => {
-                if request == closure_cid2 {
-                    network_client.respond_file(res_copy.clone(), channel).await;
-                }
-            }
-            e => todo!("{:?}", e),
+            //
+        }
+        Argument::Provide { wasm, fun, args } => {
+            let ipfs = IpfsClient::default();
+
+            // Pull Wasm *out* of IPFS
+            let wasm_bytes = ipfs
+                .cat(&wasm.as_str())
+                .map_ok(|chunk| chunk.to_vec())
+                .try_concat()
+                .await
+                .map(String::from_utf8)??;
+
+            let wasm_args = future::try_join_all(args.iter().map(|arg| async {
+                // Pull arg *out* of IPFS
+                let vec = ipfs
+                    .cat(arg.as_str())
+                    .map_ok(|chunk| chunk.to_vec())
+                    .try_concat()
+                    .await
+                    .expect("To grab arguments.");
+
+                i32::from_str(str::from_utf8(&vec).expect("Valid Wasm value."))
+                    .map(wasmer::Value::I32)
+            }))
+            .await?;
+
+            let mut store = Store::default();
+            let module = Module::new(&store, wasm_bytes).expect("Wasm module to export");
+
+            let imports = imports! {};
+            let instance =
+                Instance::new(&mut store, &module, &imports).expect("Wasm instance to be here");
+
+            let imports = imports! {};
+            let instance =
+                Instance::new(&mut store, &module, &imports).expect("Wasm instance to be here");
+
+            let _function = instance
+                .exports
+                .get::<Function>(&fun.as_str())
+                .expect("a Wasm function");
+            let _types: &[Type] = _function.ty(&store).params();
+
+            // FIXME write Wasm::Value -> Ipld converter
+            let boxed_results: Box<[Value]> = _function
+                .call(&mut store, wasm_args.as_slice())
+                .expect("tried to call function");
+
+            let res: String = boxed_results
+                .into_vec()
+                .iter()
+                .map(|v| v.to_string())
+                .collect();
+
+            let resource = Url::parse(format!("ipfs://{}", wasm).as_str()).expect("IPFS URL");
+
+            let ipld_args: Ipld = Ipld::List(
+                wasm_args
+                    .iter()
+                    .map(|a| match a {
+                        Value::I32(i) => Ipld::from(*i),
+                        _ => todo!(),
+                    })
+                    .collect(),
+            );
+
+            let closure = Closure {
+                resource,
+                action: Action::try_from("wasm/run")?,
```

Comment:
Why `try_from`? AFAICT the smart constructor should be unfailable, no?

Reply:
b/c this is fromstr under the hood, and the canonical approach seems to be tryfrom for this (and use fromstr at the bottom)

Reply:
Huh, okay. Seems wasteful, but sure 🤷‍♀️

Reply:
actually, yeah, the normal From is fine (i was pre-optimizing). will return.